### PR TITLE
raftstore: fix stale peer check

### PR DESCRIPTION
**Pull request by BusyJay**
If a TiKV is inactive for a while (block on IO for instance), all the replicas on the store may be removed by pd. So when the inactive monitor checks pd and finds out that a replica is removed, it will schedule GC message to raftstore. However, if the replica is applying snapshot at that time, TiKV will panic because a peer in applying state shouldn't be removed for safety.

This pr fixes it by checking if a peer is applying snapshot before accepting GC message.

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -672,31 +672,18 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-                async_remove = initialized;
-                if p.is_applying_snapshot() {
-                    if !p.mut_store().cancel_applying_snap() {
-                        info!(
-                            "[region {}] Stale peer {} is applying snapshot, will destroy next \
-                             time.",
-                            region_id,
-                            p.peer_id()
-                        );
-                        self.raft_metrics.message_dropped.stale_peer += 1;
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: do we not need this metric now?

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -986,32 +962,28 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-                    "[region {}] peer {:?} receives gc message, remove",
+                    "[region {}] peer {:?} receives gc message, trying to remove",
                     region_id,
                     msg.get_to_peer()
                 );
-                need_remove = true;
-                peer.pending_remove = true;
-                async_remove = peer.get_store().is_initialized();
+                job = peer.maybe_destroy();
+                if job.is_none() {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: if the job is not None, do we still need to increase stale_peer counter?

Comment (BusyJay) [reply src/raftstore/store/store.rs]: No, the metrics is used for counting dropped messages.

**Comment by siddontang** [comment]
LGTM

PTAL @overvenus

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -672,31 +672,18 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+                    None => {
+                        self.raft_metrics.message_dropped.applying_snap += 1;
                         return Ok(false);
                     }
-                    // There is no tasks in apply worker.
-                    async_remove = false;
-                }
-                p.pending_remove = true;
-                stale_peer = Some(p.peer.clone());
+                    Some(job) => Some((job, p.peer.clone())),
```
Comment (overvenus) [inline src/raftstore/store/store.rs]: Why not pass a peer reference?

File: src/raftstore/store/store.rs [hunk 4]
```diff
@@ -986,32 +962,28 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
                     msg.get_to_peer()
                 );
-                need_remove = true;
-                peer.pending_remove = true;
-                async_remove = peer.get_store().is_initialized();
+                job = peer.maybe_destroy();
+                if job.is_none() {
+                    self.raft_metrics.message_dropped.applying_snap += 1;
+                    return;
+                }
```
Comment (overvenus) [inline src/raftstore/store/store.rs]: Why not call `job.execute(self)` at here?

File: src/raftstore/store/store.rs (hunk 3 above)
Comment (BusyJay) [reply src/raftstore/store/store.rs]: When job is executed, `self` should not be borrowed.

File: src/raftstore/store/store.rs (hunk 4 above)
Comment (BusyJay) [reply src/raftstore/store/store.rs]: Same as above.

**Comment by siddontang** [comment]
PTAL @overvenus

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**